packages/websocket/src/index.ts

```diff
@@ -1,72 +1,199 @@
+ */
+export const makeReconnectingWS = (
+  url: string,
+  protocols?: string | string[],
+  options: WSReconnectOptions = {}
+) => {
+  let retries = options.retries || Infinity;
+  let ws: ReconnectingWebSocket;
+  const queue: WSMessage[] = [];
+  let events: Parameters<WebSocket["addEventListener"]>[] = [
+    ["close", () => {
+      retries-- > 0 && setTimeout(getWS, options.delay || 3000)
+    }]
+  ];
+  const getWS = () => {
+    if (ws && ws.readyState < 2) ws.close();
+    ws = Object.assign(makeWS(url, protocols, queue), {
+      reconnect: getWS
+    });
+    events.forEach(args => ws.addEventListener(...args));
+  }
+  getWS();
+  const wws: Partial<ReconnectingWebSocket> = {
+    close: (...args: Parameters<WebSocket["close"]>) => { 
+      retries = 0;
+      return ws.close(...args);
+    },
+    addEventListener: (...args: Parameters<WebSocket["addEventListener"]>) => {
+      events.push(args);
+      return ws.addEventListener(...args);
+    },
+    removeEventListener: (...args: Parameters<WebSocket["removeEventListener"]>) => {
+      events = events.filter((ev) => args[0] !== ev[0] || args[1] !== ev[1]);
+      return ws.removeEventListener(...args);
+    },
+    send: (msg: WSMessage) => {
+      wws.send!.before?.();
+      return ws.send(msg);
     }
-  };
-  // Connect the socket to the server
-  const connect = () => {
-    cancelReconnect();
-    setState(WebSocket.CONNECTING);
-    socket = new WebSocket(url, protocols);
-    socket.onopen = () => setState(WebSocket.OPEN);
-    socket.onclose = () => {
-      setState(WebSocket.CLOSED);
-      if (reconnectLimit && reconnectLimit > reconnections) {
-        reconnections += 1;
-        reconnectId = setTimeout(connect, reconnectInterval);
-      }
-    };
-    socket.onerror = onError;
-    socket.onmessage = onData;
-  };
-  onCleanup(() => disconnect);
-  return [connect, disconnect, send, state, () => socket!];
+  }
+  for (let name in ws!)
+    wws[name as keyof typeof wws] == null &&
+      Object.defineProperty(wws, name, {
+        enumerable: true,
+        get: () => typeof ws[name as keyof typeof ws] === 'function'
+          ? (ws[name as keyof typeof ws] as Function).bind(ws)
+          : ws[name as keyof typeof ws]
+      });
+  return wws as ReconnectingWebSocket;
 };
 
-export default createWebsocket;
+/**
+ * Returns a WebSocket-like object that under the hood opens new connections on disconnect and closes on cleanup:
+ * ```ts
+ * const ws = makeReconnectingWS("ws:localhost:5000");
+ * createEffect(() => ws.send(serverMessage()));
+ * ```
+ * Will not throw if you attempt to send messages before the connection opened; instead, it will enqueue the message to be sent when the connection opens.
+ */
+export const createReconnectingWS = (
+  url: string,
+  protocols?: string | string[],
+  options?: WSReconnectOptions
+) => {
+  const ws = makeReconnectingWS(url, protocols, options);
+  onCleanup(() => ws.close());
+  return ws;
+}
```

Comment:
I like the strong make vs create separation 💪
btw you can type it using `typeof makeReconnectingWS` if youre making the exact same interface. Best way to ensure they are in sync
```suggestion
export const createReconnectingWS: typeof makeReconnectingWS = (url, protocols, options) => {
  const ws = makeReconnectingWS(url, protocols, options);
  onCleanup(() => ws.close());
  return ws;
}
```